services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/utils/ResumableFileDownloadSerializer.java

```diff
@@ -0,0 +1,269 @@
+        }
+
+        if (downloadNodes.get("s3ObjectLastModified") != null) {
+            Instant s3ObjectLastModified = instantUnmarshaller.unmarshall(context,
+                                                                          downloadNodes.get("s3ObjectLastModified"),
+                                                                          (SdkField<Instant>) instantSdkField());
+            builder.s3ObjectLastModified(s3ObjectLastModified);
+        }
+        builder.downloadFileRequest(parseDownloadFileRequest(downloadNodes.get("downloadFileRequest"),
+                                                             context, unmarshallerRegistry));
+
+        return builder.build();
+    }
+
+    private static DownloadFileRequest parseDownloadFileRequest(JsonNode fileRequest,
+                                                                JsonUnmarshallerContext context,
+                                                                JsonUnmarshallerRegistry unmarshallerRegistry) {
+        DownloadFileRequest.Builder fileRequestBuilder = DownloadFileRequest.builder();
+        Map<String, JsonNode> fileRequestNodes = fileRequest.asObject();
+
+        fileRequestBuilder.destination(Paths.get(fileRequestNodes.get("destination").asString()));
+
+        GetObjectRequest.Builder getObjectBuilder = GetObjectRequest.builder();
+        JsonNode getObjectRequest = fileRequestNodes.get("getObjectRequest");
+        Map<String, JsonNode> getObjectRequestNodes = getObjectRequest.asObject();
+        getObjectRequestNodes.forEach((key, value) -> {
+            SdkField<?> f = getSdkField(key);
+            MarshallingType<?> marshallingType = f.marshallingType();
+            if (marshallingType.equals(MarshallingType.STRING)) {
+                f.set(getObjectBuilder, value.asString());
+            } else if (marshallingType.equals(MarshallingType.INSTANT)) {
+                JsonUnmarshaller<Object> instantUnmarshaller =
+                    unmarshallerRegistry.getUnmarshaller(MarshallLocation.PAYLOAD, MarshallingType.INSTANT);
+                Instant instantValue = (Instant) instantUnmarshaller.unmarshall(context, value,
+                                                                                (SdkField<Object>) instantSdkField());
+                f.set(getObjectBuilder, instantValue);
+            } else if (marshallingType.equals(MarshallingType.INTEGER)) {
+                JsonUnmarshaller<Object> integerUnmarshaller =
+                    unmarshallerRegistry.getUnmarshaller(MarshallLocation.PAYLOAD, MarshallingType.INTEGER);
+                Integer integerValue = (Integer) integerUnmarshaller.unmarshall(context, value, (SdkField<Object>) f);
+                f.set(getObjectBuilder, integerValue);
+            } else {
+                throw new IllegalStateException("should not happen");
+            }
```

Comment:
Would it make sense to use switch?

Reply:
It's not a switch compatible type, so I'd have to convert to string which in this case seemed awkward but will look into it with the copy rewrite.